day-11/part-2/badouralix.rs

```diff
@@ -0,0 +1,199 @@
+                        .collect::<String>()
+                        .parse()
+                        .unwrap(),
+                )
+            } else if operation.contains('*') {
+                test_op_mul = Some(
+                    operation
+                        .chars()
+                        .skip(25)
+                        .collect::<String>()
+                        .parse()
+                        .unwrap(),
+                )
+            } else {
+                unimplemented!()
+            }
+        }
+
+        let test_divisible_by = lines
+            .next()
+            .unwrap()
+            .chars()
+            .skip(21)
+            .collect::<String>()
+            .parse()
+            .unwrap();
+
+        let test_dst_monkey_true = lines.next().unwrap().as_bytes()[29] as usize - b'0' as usize;
+
+        let test_dst_monkey_false = lines.next().unwrap().as_bytes()[30] as usize - b'0' as usize;
+
+        Monkey {
+            id,
+            items,
+            test_divisible_by,
+            test_dst_monkey_false,
+            test_dst_monkey_true,
+            test_op_add,
+            test_op_mul,
+            test_op_square,
+        }
+    }
+
+    fn throw(&self, old: isize, ppcm: isize) -> (isize, usize) {
+        let new = if let Some(v) = self.test_op_add {
+            (old + v) % ppcm
+        } else if let Some(v) = self.test_op_mul {
+            (old * v) % ppcm
+        } else if self.test_op_square {
+            (old * old) % ppcm
+        } else {
+            unimplemented!()
+        };
+
+        if new % self.test_divisible_by == 0 {
+            (new, self.test_dst_monkey_true)
+        } else {
+            (new, self.test_dst_monkey_false)
+        }
+    }
+}
+
+impl Display for Monkey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Monkey {}: {:?}", self.id, self.items)
+    }
+}
+
+fn main() {
+    let now = Instant::now();
+    let output = run(&args().nth(1).expect("Please provide an input"));
+    let elapsed = now.elapsed();
+    println!("_duration:{}", elapsed.as_secs_f64() * 1000.);
+    println!("{}", output);
+}
+
+fn run(input: &str) -> usize {
+    let mut inspects = Vec::new();
+    let mut monkeys = Vec::new();
+    for (idx, definition) in input.split("\n\n").enumerate() {
+        assert!(definition.starts_with(&format!("Monkey {idx}:\n")));
+        inspects.push(0);
+        monkeys.push(Monkey::new(definition));
+    }
+
+    let mut ppcm = 1;
+    for monkey in &monkeys {
+        ppcm *= monkey.test_divisible_by;
+    }
+
+    for _ in 0..10000 {
+        for src in 0..monkeys.len() {
+            let items = monkeys[src].items.to_owned();
+            for item in items {
+                let (new, dst) = monkeys[src].throw(item, ppcm);
+                monkeys[dst].items.push(new);
+            }
+            inspects[src] += monkeys[src].items.len();
+            monkeys[src].items = Vec::new();
```

Comment:
```suggestion
            monkeys[src].items.clear();
```

Reply:
Oh wow removing this extra allocation cuts the run time in half !